### ISSUE TITLE
Docker event documentation incorrect

### ISSUE THREAD

**Issue by samuelkarp** [docker/docker#21331]
Hello,

It looks like the documentation for [events](https://docs.docker.com/engine/reference/api/docker_remote_api/#docker-events) has become out of date, in particular around the `kill` event.  Prior to Docker 1.7.0, the `kill` event was only emitted on `SIGKILL` being sent to the container via `docker kill`.  Since 1.7.0, the semantic changed and the `kill` event started to be emitted on any signal sent via `docker kill`, and since 1.8.0 the semantic changed again and the `kill` event started to be emitted for any signal sent at all (including `SIGTERM` and `SIGKILL` sent as part of `docker stop`).  These changes in behavior led to a [bug](https://github.com/aws/amazon-ecs-agent/issues/335) with the Amazon ECS Agent, which we will be correcting with our next release.

Here's the set of behavior changes I've found across different Docker versions:

| Docker version | `docker kill -s USR1` | `docker kill`  | `docker stop`                  | Container exit |
|----------------|-----------------------|----------------|--------------------------------|----------------|
| 1.5.0          | no event              | `die`, `kill`  | `die`, `stop`                  | `die`          |
| 1.6.0          | no event              | `die`, `kill`  | `die`, `stop`                  | `die`          |
| 1.6.1          | no event              | `die`, `kill`  | `die`, `stop`                  | `die`          |
| 1.6.2          | no event              | `die`, `kill`  | `die`, `stop`                  | `die`          |
| 1.7.0          | `kill`                | `die`, `kill`  | `die`, `stop`                  | `die`          |
| 1.7.1          | `kill`                | `die`, `kill`  | `die`, `stop`                  | `die`          |
| 1.8.0          | `kill`                | `kill`, `die`  | `kill`, `kill`, `die`, `stop`  | `die`          |
| 1.8.1          | `kill`                | `kill`, `die`  | `kill`, `kill`, `die`, `stop`  | `die`          |
| 1.8.2          | `kill`                | `kill`, `die`  | `kill`, `kill`, `die`, `stop`  | `die`          |
| 1.8.3          | `kill`                | `kill`, `die`  | `kill`, `kill`, `die`, `stop`  | `die`          |
| 1.9.0          | `kill`                | `kill`, `die`  | `kill`, `kill`, `die`, `stop`  | `die`          |
| 1.9.1          | `kill`                | `kill`, `die`  | `kill`, `kill`, `die`, `stop`  | `die`          |
| 1.10.0         | `kill`*               | `kill`, `die`* | `kill`, `kill`, `die`, `stop`* | `die`*         |
| 1.10.1         | `kill`*               | `kill`, `die`* | `kill`, `kill`, `die`, `stop`* | `die`*         |
| 1.10.2         | `kill`*               | `kill`, `die`* | `kill`, `kill`, `die`, `stop`* | `die`*         |
| 1.10.3         | `kill`*               | `kill`, `die`* | `kill`, `kill`, `die`, `stop`* | `die`*         |

\* These events are now typed with `container`.  Other events are emitted with type `network` and `volume`.

Note that `kill` is emitted under many more circumstances in recent versions of Docker, and that the ordering of `die`, `kill` changed to `kill`, `die` with Docker 1.8.0 as well.  The [available](https://docs.docker.com/engine/reference/commandline/events/) [documentation](https://docs.docker.com/engine/reference/api/docker_remote_api/#docker-events) for events is fairly light as a whole, but the [event diagram](https://docs.docker.com/engine/reference/api/images/event_state.png) has been out of date at least since 1.7.0.

It would be fantastic if the semantics of the different events were documented and any semantic changes were called out in the changelog for releases.

Thanks!
Sam

**Comment by thaJeztah** [comment]
Hi @samuelkarp - going through old issues; would you be interested in opening a PR if you think this still needs changes in the docs?

**Comment by samuelkarp** [comment]
@thaJeztah Yes, I think this still needs changes as the docs are still incorrect with respect to the `kill` event.  I'd be happy to open a PR, but I don't know when I'm going to have the time to do so (especially since I don't have the source of the current diagram in the docs).  I'd also be interested in your thoughts on how you'd address the semantics of events changing over time (e.g., which events are emitted under which situation and how they change in different Docker versions).

**Comment by thaJeztah** [comment]
The source of the graphic is actually in the repo; https://github.com/docker/docker/blob/master/docs/reference/api/images/event_state.gliffy, which can be used on https://www.gliffy.com :smile:

**Comment by samuelkarp** [comment]
Oh neat!  Now the bigger question is finding time.  If anyone else wants to pick this up before I've had the chance, please do!

**Comment by thaJeztah** [comment]
:smile:

**Comment by dccurtis** [comment]
I can update the docs but it looks like these links are out of date.

The events documentation @samuelkarp mentioned redirects to https://docs.docker.com/engine/api/ and the graphic event_state.gliffy no longer exists.

Is this still an issue?  If so, can someone point me to the docs that need to be updated?

**Comment by samuelkarp** [comment]
@dccurtis It looks like the event documentation was moved [here](https://docs.docker.com/engine/api/v1.26/#operation/SystemEvents), though now it contains no lifecycle information about container events.  I don't know if that was an intentional omission or not.

**Comment by thaJeztah** [comment]
I'll have to ask @bfirsh @mstanleyjones if that was deliberate or just got lost in the process

**Comment by dccurtis** [comment]
@thaJeztah Sounds good.  I'll stay tuned to see if I can help out with this.

**Comment by bfirsh** [comment]
Probably lost in the process. At least, I don't remember removing it intentionally. Sorry about that!

**Comment by mstanleyjones** [comment]
🤷‍♀️
